<?xml version="1.0"?>
<launch>
  <arg name="model" default="$(find archileon_ros_t1)/urdf/alm_t1.urdf"/>
  <arg name="gui" default="true" />
  <param name="robot_description" command="$(find xacro)/xacro $(arg model)" />
  <param name="use_gui" value="$(arg gui)"/>
  <arg name="rvizconfig" default="$(find archileon_ros_t1)/config/SETTING.rviz" />
  <arg name="vive" default="false"/>

  <node name="joy_node" pkg="joy" type="joy_node" >
    <param name="coalesce_interval" value="0.1"/>
  </node>

  <node name="move_driver" pkg="archileon_ros_t1" type="move_driver">
    <remap from="servo0" to="pif/FL/servo1"/>
    <remap from="servo1" to="pif/FR/servo1"/>
    <remap from="servo2" to="pif/BL/servo1"/>
    <remap from="servo3" to="pif/BR/servo1"/>

    <remap from="servo4" to="pif/BL/servo2"/>
    <remap from="servo5" to="pif/BR/servo2"/>

    <remap from="motor0" to="pif/FL/motor"/>
    <remap from="motor1" to="pif/FR/motor"/>
    <remap from="motor2" to="pif/BL/motor"/>
    <remap from="motor3" to="pif/BR/motor"/>
  </node>

  <group unless="$(arg vive)">
    <node name="move_sim" pkg="archileon_ros_t1" type="move_sim">
      <remap from="steer0" to="pif/FL/servo1"/>
      <remap from="steer1" to="pif/FR/servo1"/>
      <remap from="steer2" to="pif/BL/servo1"/>
      <remap from="steer3" to="pif/BR/servo1"/>

      <remap from="steer4" to="pif/BL/servo2"/>
      <remap from="steer5" to="pif/BR/servo2"/>

      <remap from="motor0" to="pif/FL/motor"/>
      <remap from="motor1" to="pif/FR/motor"/>
      <remap from="motor2" to="pif/BL/motor"/>
      <remap from="motor3" to="pif/BR/motor"/>
    </node>
  </group>

  <node name="pif_to_joint" pkg="archileon_ros_t1" type="pif_to_joint">
    <remap from="steer0" to="pif/FL/servo1"/>
    <remap from="steer1" to="pif/FR/servo1"/>
    <remap from="steer2" to="pif/BL/servo1"/>
    <remap from="steer3" to="pif/BR/servo1"/>
    <remap from="motor0" to="pif/FL/motor"/>
    <remap from="motor1" to="pif/FR/motor"/>
    <remap from="motor2" to="pif/BL/motor"/>
    <remap from="motor3" to="pif/BR/motor"/>
    <remap from="tube0"  to="pif/BL/servo2"/>
    <remap from="tube1"  to="pif/BR/servo2"/>
  </node>

  <!-- <group ns="vive" unless="$(arg vive)">  -->
  <node name="test_tf_listen" pkg="archileon_ros_t1" type="test_tf_listen" />
  <node name="tf_bl_rot_listen" pkg="archileon_ros_t1" type="tf_bl_rot_listen" />
  <node name="tf_br_rot_listen" pkg="archileon_ros_t1" type="tf_br_rot_listen" />

  <!-- </group> -->

  <group ns="vive" if="$(arg vive)"> 
    <node pkg="tf" type="static_transform_publisher" name="vive_world_orientation2" args="0.0 0.0 1.55 0 0.0 0 world map 100" />
    <node name="vive_world" pkg="archileon_ros_t1" type="vive_world.py" output="screen"/>
    <node name="vive_tracker" pkg="archileon_ros_t1" type="vive_tracker.py" output="screen"/>
    <!-- connect vive to robot--> 
    <!-- Be careful about the used vive tracker' serial number for the used frame like LHR_!!!!!!-->
    <node pkg="tf" type="static_transform_publisher" name="vive_robot" args="0.185 0.0 -0.24 0 0.0 0 LHR_1CDCEA0B base_link 100" /> 
  </group>



<!-- Visualizer -->
  <node name="map_visualizer" pkg="archileon_ros_t1" type="map_visualizer" />
  <node name="map_way_points_publisher" pkg="archileon_ros_t1" type="map_way_points_publisher.py" />

  <!-- <node name="map_trajectory_visualizer" pkg="archileon_ros_t1" type="map_trajectory_visualizer.py" /> -->
  <node name="map_target_visualizer" pkg="archileon_ros_t1" type="map_target_visualizer.py"/>
  

  <node name="rviz" pkg="rviz" type="rviz" args="-d $(arg rvizconfig)" />
  <node name="robot_state_publisher" pkg="robot_state_publisher" type="state_publisher" />

  <node name="FL_servo1" pkg="archileon_ros_t1" type="pif_servo1">
    <param name="ID" value="FL"/>
    <param name="REDUCTION" value="-1.0"/>
    <remap from="servo" to="pif/FL/servo1"/>
  </node>
  <node name="FR_servo1" pkg="archileon_ros_t1" type="pif_servo1">
    <param name="ID" value="FR"/>
    <param name="REDUCTION" value="-1.0"/>
    <remap from="servo" to="pif/FR/servo1"/>
  </node>
  <node name="BL_servo1" pkg="archileon_ros_t1" type="pif_servo1">
    <param name="ID" value="BL"/>
    <param name="REDUCTION" value="-1.0"/>
    <remap from="servo" to="pif/BL/servo1"/>
  </node>
  <node name="BR_servo1" pkg="archileon_ros_t1" type="pif_servo1">
    <param name="ID" value="BR"/>
    <param name="REDUCTION" value="-1.0"/>
    <remap from="servo" to="pif/BR/servo1"/>
  </node>

  <node name="BL_servo2" pkg="archileon_ros_t1" type="pif_servo2">
    <param name="ID" value="BL"/>
    <param name="REDUCTION" value="1.2"/>
    <remap from="servo" to="pif/BL/servo2"/>
  </node>
  <node name="BR_servo2" pkg="archileon_ros_t1" type="pif_servo2">
    <param name="ID" value="BR"/>
    <param name="REDUCTION" value="1.2"/>
    <remap from="servo" to="pif/BR/servo2"/>
  </node>

  <node name="FL_motor" pkg="archileon_ros_t1" type="pif_motor">
    <param name="ID" value="FL"/>
    <param name="ACC" value="-2.0"/>
    <remap from="motor" to="pif/FL/motor"/>
  </node>
  <node name="FR_motor" pkg="archileon_ros_t1" type="pif_motor">
    <param name="ID" value="FR"/>
    <param name="ACC" value="2.0"/>
    <remap from="motor" to="pif/FR/motor"/>
  </node>
  <node name="BL_motor" pkg="archileon_ros_t1" type="pif_motor">
    <param name="ID" value="BL"/>
    <param name="ACC" value="-2.0"/>
    <remap from="motor" to="pif/BL/motor"/>
  </node>
  <node name="BR_motor" pkg="archileon_ros_t1" type="pif_motor">
    <param name="ID" value="BR"/>
    <param name="ACC" value="2.0"/>
    <remap from="motor" to="pif/BR/motor"/>
  </node>

  <node name="arduino_serial" pkg="archileon_ros_t1" type="arduino_serial" output="screen"/>
  <node name="autonomous_operation" pkg="archileon_ros_t1" type="autonomous_operation.py" output="screen"/>
  <node name="tube_operation" pkg="archileon_ros_t1" type="tube_operation.py" output="screen"/>
  <node name="showRobWaypoints" pkg="archileon_ros_t1" type="showRobWaypoints.py" />



</launch>
